cache/lru_cache.cc

```diff
@@ -643,37 +647,131 @@ std::string LRUCacheShard::GetPrintableOptions() const {
-        /* max_upper_hash_bits */ 32 - num_shard_bits, secondary_cache);
+namespace {
+#ifndef ROCKSDB_LITE
+static std::unordered_map<std::string, CacheMetadataChargePolicy>
+    metadata_charge_policy_string_map = {
+        {"kDontCharge", CacheMetadataChargePolicy::kDontChargeCacheMetadata},
+        {"kFullCharge", CacheMetadataChargePolicy::kFullChargeCacheMetadata},
+};
+#endif  // ROCKSDB_LITE
+
+static std::unordered_map<std::string, OptionTypeInfo>
+    lru_cache_options_type_info = {
+#ifndef ROCKSDB_LITE
+        {"capacity",
+         {offsetof(struct LRUCacheOptions, capacity), OptionType::kSizeT,
+          OptionVerificationType::kNormal, OptionTypeFlags::kNone}},
+        {"num_shard_bits",
+         {offsetof(struct LRUCacheOptions, num_shard_bits), OptionType::kInt,
+          OptionVerificationType::kNormal, OptionTypeFlags::kNone}},
+        {"strict_capacity_limit",
+         {offsetof(struct LRUCacheOptions, strict_capacity_limit),
+          OptionType::kBoolean, OptionVerificationType::kNormal,
+          OptionTypeFlags::kNone}},
+        {"metadata_charge_policy",
+         OptionTypeInfo::Enum(
+             offsetof(struct LRUCacheOptions, metadata_charge_policy),
+             &metadata_charge_policy_string_map)},
+        {"high_pri_pool_ratio",
+         {offsetof(struct LRUCacheOptions, high_pri_pool_ratio),
+          OptionType::kDouble, OptionVerificationType::kNormal,
+          OptionTypeFlags::kMutable}},
+        {"secondary_cache",
+         OptionTypeInfo::AsCustomSharedPtr<SecondaryCache>(
+             offsetof(struct LRUCacheOptions, secondary_cache),
+             OptionVerificationType::kNormal,
+             OptionTypeFlags::kAllowNull | OptionTypeFlags::kCompareNever)},
+#endif  // ROCKSDB_LITE
+};
+}  // namespace
+LRUCache::LRUCache(const LRUCacheOptions& options)
+    : ShardedCache(), shards_(nullptr), options_(options) {
+  RegisterOptions(&options_, &lru_cache_options_type_info);
+}
+
+LRUCache::LRUCache() : ShardedCache(), shards_(nullptr) {
+  RegisterOptions(&options_, &lru_cache_options_type_info);
+}
+
+bool LRUCache::IsMutable() const {
+  MutexLock l(&capacity_mutex_);
+  return (shards_ == nullptr);
+}
+
+Status LRUCache::PrepareOptions(const ConfigOptions& config_options) {
+  MutexLock l(&capacity_mutex_);
+  if (shards_ != nullptr) {  // Already prepared
+    return Status::OK();
+  } else if (options_.high_pri_pool_ratio < 0.0 ||
+             options_.high_pri_pool_ratio > 1.0) {
+    // invalid high_pri_pool_ratio
+    return Status::InvalidArgument(
+        "LRUCache: High priority pool ratio out of bounds");
+  } else if (options_.num_shard_bits >= 20) {
+    return Status::InvalidArgument(
+        "The cache cannot be sharded into too many fine pieces");
+  } else if (options_.num_shard_bits < 0) {
+    options_.num_shard_bits = GetDefaultCacheShardBits(options_.capacity);
+  }
+  Status s = ShardedCache::PrepareOptions(config_options);
+  if (s.ok()) {
```

Comment:
Start from here, I do not think it belongs to the function of  "PrepareOptions". We may need another function for initiation and call it after "PrepareOptions" in "NewLRUCache"